Source/Core/Core/HW/GCMemcard/GCMemcardUtils.cpp

```diff
@@ -53,4 +71,282 @@ bool HasSameIdentity(const DEntry& lhs, const DEntry& rhs)
+
+  const size_t block_count = savefile.dir_entry.m_block_count;
+  const u64 expected_size = DENTRY_SIZE + block_count * BLOCK_SIZE;
+  if (expected_size != filesize)
+    return ReadSavefileErrorCode::DataCorrupted;
+
+  if (!ReadBlocksFromIOFile(file, savefile.blocks, block_count))
+    return ReadSavefileErrorCode::IOError;
+
+  return savefile;
+}
+
+static std::variant<ReadSavefileErrorCode, Savefile> ReadSavefileInternalGCS(File::IOFile& file,
+                                                                             u64 filesize)
+{
+  std::array<u8, GCS_HEADER_SIZE> gcs_header;
+  if (!file.ReadBytes(gcs_header.data(), gcs_header.size()))
+    return ReadSavefileErrorCode::IOError;
+
+  if (std::memcmp(gcs_header.data(), GCS_MAGIC.data(), GCS_MAGIC.size()) != 0)
+    return ReadSavefileErrorCode::DataCorrupted;
+
+  Savefile savefile;
+  std::memcpy(&savefile.dir_entry, &gcs_header[GCS_DENTRY_OFFSET], DENTRY_SIZE);
+
+  // field containing the Block count as displayed within
+  // the GameSaves software is not stored in the GCS file.
+  // It is stored only within the corresponding GSV file.
+  // If the GCS file is added without using the GameSaves software,
+  // the value stored is always "1"
+
+  // to get the actual block count calculate backwards from the filesize
+  const u64 total_block_size = filesize - GCS_HEADER_SIZE;
+  if ((total_block_size % BLOCK_SIZE) != 0)
+    return ReadSavefileErrorCode::DataCorrupted;
+
+  const size_t block_count = total_block_size / BLOCK_SIZE;
+  savefile.dir_entry.m_block_count = static_cast<u16>(block_count);
+
+  if (!ReadBlocksFromIOFile(file, savefile.blocks, block_count))
+    return ReadSavefileErrorCode::IOError;
+
+  return savefile;
+}
+
+static std::variant<ReadSavefileErrorCode, Savefile> ReadSavefileInternalSAV(File::IOFile& file,
+                                                                             u64 filesize)
+{
+  std::array<u8, SAV_HEADER_SIZE> sav_header;
+  if (!file.ReadBytes(sav_header.data(), sav_header.size()))
+    return ReadSavefileErrorCode::IOError;
+
+  if (std::memcmp(sav_header.data(), SAV_MAGIC.data(), SAV_MAGIC.size()) != 0)
+    return ReadSavefileErrorCode::DataCorrupted;
+
+  Savefile savefile;
+  savefile.dir_entry = ExtractDEntryFromSavHeader(sav_header);
+
+  const size_t block_count = savefile.dir_entry.m_block_count;
+  const u64 expected_size = SAV_HEADER_SIZE + block_count * BLOCK_SIZE;
+  if (expected_size != filesize)
+    return ReadSavefileErrorCode::DataCorrupted;
+
+  if (!ReadBlocksFromIOFile(file, savefile.blocks, block_count))
+    return ReadSavefileErrorCode::IOError;
+
+  return savefile;
+}
+
+std::variant<ReadSavefileErrorCode, Savefile> ReadSavefile(const std::string& filename)
+{
+  File::IOFile file(filename, "rb");
+  if (!file)
+    return ReadSavefileErrorCode::OpenFileFail;
+
+  if (!file.Seek(0, SEEK_END))
+    return ReadSavefileErrorCode::IOError;
+  const u64 filesize = file.Tell();
```

Comment:
Could this be moved down near its usage?

Reply:
No, but what we can do is just use file.GetSize() instead.